### ISSUE TITLE
torch.no_grad() creates memory leak because of forward hook

### ISSUE THREAD

**Issue by avivt** [pytorch/opacus#176]
## 🐛 Bug

When evaluating a model within a 'with torch.no_grad():' block, when the model is still in training mode, the forward hooks will save the activations and they will not be deleted.

Setting model.eval() will fix the issue.

**Comment by alexandresablayrolles** [comment]
Hi @avivt,
Thanks for raising this issue. If I understand correctly, setting the model to eval fixed your problem? We will try to make the forward hook not store activations when in no_grad model.
